lib/Graphics/Image.js

```diff
@@ -38,466 +40,767 @@ class Image {
 				case 'right':
-					xouter = width - png.width
-					xinner = 0
-					wouter = png.width
+					destination.x += width - scaledImageWidth
 					break
 			}
 		}
 
-		if (png.height > height) {
+		if (scaledImageHeight > height) {
 			// image is taller than drawing pane
+			source.h = height / calculatedScale
 			switch (valign) {
-				case 'top':
-					youter = 0
-					yinner = 0
-					houter = height
-					break
 				case 'center':
-					youter = 0
-					yinner = Math.round((png.height - height) / 2, 0)
-					houter = height
+					source.y = (imageHeight - source.h) / 2
 					break
 				case 'bottom':
-					youter = 0
-					yinner = png.height - height
-					houter = height
+					source.y = imageHeight - source.h
 					break
 			}
-		} else {
+		} else if (scaledImageHeight < height) {
 			// image is smaller than drawing pane
+			destination.h = scaledImageHeight
 			switch (valign) {
-				case 'top':
-					youter = 0
-					yinner = 0
-					houter = png.height
-					break
 				case 'center':
-					youter = Math.round((height - png.height) / 2, 0)
-					yinner = 0
-					houter = png.height
+					destination.y += (height - scaledImageHeight) / 2
 					break
 				case 'bottom':
-					youter = height - png.height
-					yinner = 0
-					houter = png.height
+					destination.y += height - scaledImageHeight
 					break
 			}
 		}
 
-		for (let y = 0; y < houter; y++) {
-			for (let x = 0; x < wouter; x++) {
-				let idx = (png.width * (y + yinner) + x + xinner) << 2
-				let r = png.data[idx]
-				let g = png.data[idx + 1]
-				let b = png.data[idx + 2]
-				let a = png.data[idx + 3]
+		this.context2d.drawImage(
+			imageCanvas,
+			source.x,
+			source.y,
+			source.w,
+			source.h,
+			destination.x,
+			destination.y,
+			destination.w,
+			destination.h
+		)
+		//console.log('scale, image w, image h', calculatedScale, imageCanvas.width, imageCanvas.height, '->', scaledImageWidth, scaledImageHeight)
+		//console.log('drawing image ', source.x, source.y, source.w, source.h, '->', destination.x, destination.y, destination.w, destination.h)
 
-				if (a > 0) {
-					if (a === 255) {
-						this.pixel(xStart + xouter + x, yStart + youter + y, rgb(r, g, b))
-					} else {
-						this.pixel(xStart + xouter + x, yStart + youter + y, argb(a, r, g, b))
-					}
-				}
+		this.lastUpdate = Date.now()
+	}
+
+	/**
+	 * draws a single line of left aligned text
+	 * the line lenght is not wrapped or limited and may extend beyond the canvas
+	 * @param {number} x left position where to start the line
+	 * @param {number} y top position where to start the line
+	 * @param {string} text
+	 * @param {string} color CSS color string
+	 * @param {number} fontsize Em height
+	 * @param {boolean|undefined} dummy Don't actually draw anything just return the text width
+	 * @returns {number} width of the line
+	 */
+	drawTextLine(x, y, text, color, fontsize, dummy) {
+		if (text === undefined || text.length == 0) return 0
+
+		if (parseInt(fontsize) == NaN) return 0
```

Comment:
This check will always return false, as `NaN` is not equal to itself. Use `isNaN` instead